### ISSUE TITLE
Missing CAN J1939 support and vcan0 in latest docket image.

### ISSUE THREAD

**Issue by menschel** [docker-library/python#618]
Hi,

I recently found that J1939 is not compiled into the latest docker image.

https://gitlab.com/Menschel/socketcan/-/jobs/1315560259

This may be caused by not having

`linux/can/j1939.h`

available at compile time.

The above mentioned tox tests runs smoothly on a raspberry pi with natively compiled Python 3.9.5.

Last but not least, It would be great if the docker images would have vcan0 enabled by default.
I'm using vcan0 for tests and coverage in pipeline.

Thank you very much.

**Comment by tianon** [comment]
Can you provide a simple reproducer, perhaps in the form of a single `docker run` one-liner that imports a module which currently fails but will succeed if we have the appropriate header available during compilation?

**Comment by menschel** [comment]
Concerning the missing J1939 support:

I don't use docker directly but via [gitlab-ci](https://gitlab.com/Menschel/socketcan/-/blob/master/.gitlab-ci.yml)
`image: python:latest`
Reproducer would be something like this.
```
python -i
socket.J1939_NO_NAME
```
which will then raise Attribute Error

Actually it would be this [unit test](https://github.com/python/cpython/blob/main/Lib/test/test_socket.py#L2208).
The build depends on the specific [header file](https://github.com/python/cpython/blob/main/Modules/socketmodule.h#L147) include which ideally should be there after kernel version 5.4 . Raspberry Pi kernel got that header file 4 weeks ago, although on kernel version 5.10 since a few month.


Concerning the vcan0.
It is actually a workaround to test CAN related stuff in an unprivileged mode like the public gitlab runners.
it basically comes down to this.
```
sudo ip link add type vcan
sudo ip link set vcan0 up
```
I'm not sure if this really is the philosophers stone. It would just be nice to have.
My prediction is that there will be a lot of python applications targeting CAN bus in the near future because Kernel support is mature on almost every embedded platform and CAN Hardware for Raspberry Pi became dead-cheap lately.

**Comment by yosifkit** [comment]
It looks like `j1939.h` doesn't exist until Debian Bullseye which has not migrated to `stable` yet, so we don't have a python image based on it.

https://packages.debian.org/search?suite=buster&searchon=contents&keywords=j1939.h
https://packages.debian.org/search?suite=bullseye&searchon=contents&keywords=j1939.h

If python were compiled with this support, would it have issues running for users still on kernels earlier than `5.10`?

**Comment by menschel** [comment]
The safe threshold should be `>5.4.y`  . Technically kernel raises an OSError Protocol Not Supported regardless of if the constants are compiled into python or not. It cannot do harm on it's own IMO.

**Comment by menschel** [comment]
Just checked, I was active in raspbian repo 25 days ago, so I don't understand how this header file did not make it into bullseye while raspbian buster already had it.
https://github.com/raspberrypi/linux/pull/4346

**Comment by menschel** [comment]
Ok, thank you for the explanation.
Guess I'll have to set up my own runner after all.

**Comment by tianon** [comment]
Fixed by #633 :+1:
